### PR TITLE
Add OCI Language MCP Server

### PR DESCRIPTION
**Pull request by antimverma**
# Description

Adds the OCI Language MCP server under `src/oci-language-mcp-server`.

This server provides OCI Language shared pretrained capabilities through seven typed MCP tools:

- Dominant language detection
- Text classification
- Named-entity detection
- Key phrase extraction
- Sentiment analysis
- PII detection and transformation
- Text translation

The implementation follows the repository MCP server structure and public packaging expectations:

- Adds `pyproject.toml`, `uv.lock`, `README.md`, `CHANGELOG.md`, `LICENSE.txt`, `Containerfile`, `.containerignore`, and the `oracle/` package source.
- Exposes the package entry point as `oracle.oci-language-mcp-server`.
- Supports stdio and hardened Streamable HTTP transports.
- Supports outbound OCI session-token, instance-principal, and resource-principal authentication through `oracle-mcp-common`.
- Supports remote MCP caller authentication through OAuth resource-server validation or a controlled token file.
- Keeps inbound caller authentication separate from outbound OCI identity: OCI Language calls use the server’s configured OCI identity and do not exchange incoming OAuth tokens.
- Adds Host/Origin allowlists, request-body limits, remote request rate limits, bounded concurrent OCI execution, health/readiness endpoints, safe telemetry, and sanitized error responses.
- Adds OCI SDK `additional_user_agent` telemetry derived from package metadata for every supported outbound authentication mode.
- Adds typed request validation, OCI request mapping, structured result parsing, bounded human-readable output, and PII-safe transformation behavior.
- Documents all runtime settings, local and remote deployment, Docker usage, authentication behavior, IAM requirements, and local validation.
- Configures the container for a non-root, read-only runtime and builds from the repository root so the shared `src/common` workspace dependency is available.

Fixes: N/A

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

Validated locally from `src/oci-language-mcp-server`:

- [x] `pytest --cov=. --cov-branch --cov-report=term-missing`
  - `87 passed`
  - Coverage: `91.00%`
- [x] `uv lock --check`
- [x] `ruff check --no-cache .`
- [x] HTTP runtime probe
  - Verified `GET /health` returns `{"status":"ok"}`
  - Verified `GET /ready` returns `{"status":"ready"}`
- [x] Auth and provider tests cover session-token, instance-principal, and resource-principal OCI client construction paths.
- [x] Remote OAuth and token-file HTTP security behavior are covered by integration tests.
- [x] Workspace dependency resolution was validated with the repository-root container build layout.

**Test Configuration**:

- OS: macOS
- Python: 3.13.2
- Package manager: uv
- MCP transports: stdio and Streamable HTTP
- OCI authentication modes: session-token, instance principal, and resource principal

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove the feature works
- [x] New and existing unit and integration tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules (N/A)

### CODE REVIEW

**Comment by antimverma** [comment]
Hi @krisrice @dustin-sale Can you please help review/approve this PR for adding OCI Language MCP ?

File: src/oci-language-mcp-server/oracle/oci_language_mcp_server/models.py [hunk 1]
```diff
@@ -0,0 +1,532 @@
...
+            return value
+        return _validate_language_code(value)
+
+
+class ToolOptions(StrictModel):
+    region: str | None = Field(
+        default=None,
+        min_length=1,
+        max_length=64,
+        description="OCI region override for this call.",
```
Comment (dustin-sale) [inline src/oci-language-mcp-server/oracle/oci_language_mcp_server/models.py:125]: _Comment from @dustin-sale via Codex._

[P1] Restrict caller-controlled OCI regions

Evidence: `options.region` is only length-bounded and flows into `AIServiceLanguageClient`. OCI SDK 2.179.0 treats a dotted value as a domain; `attacker.example` produces `https://language.aiservice.attacker.example`.

Impact: An authenticated caller can redirect submitted documents and OCI-signed headers, including security-token identity material, to an attacker-controlled or internal HTTPS endpoint. Arbitrary values also grow the permanent region-client cache.

Requested change: Accept only trusted OCI region identifiers or an operator allowlist, keep custom endpoints operator-controlled, bound the client cache, and test dotted, userinfo, loopback, and port-bearing values.

File: src/oci-language-mcp-server/oracle/oci_language_mcp_server/provider.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+            "opc_request_id": opc_request_id,
+            "retry_strategy": oci.retry.NoneRetryStrategy(),
+        }
+
+    def _client(self, *, region: str | None) -> Any:
+        auth = build_auth_context(self._settings, region=region)
+        cache_key = (auth.mode, auth.region, self._settings.oci_service_endpoint)
+        if auth.mode != "session" and cache_key in self._clients:
+            return self._clients[cache_key]
+        config = {**auth.config, "additional_user_agent": _ADDITIONAL_UA}
```
Comment (dustin-sale) [inline src/oci-language-mcp-server/oracle/oci_language_mcp_server/provider.py]: _Comment from @dustin-sale via Codex._

[P1] Disable OCI SDK request logging

Evidence: This merge preserves a session profile's `log_requests=true`. OCI SDK 2.179.0 then enables `HTTPConnection.debuglevel=1`, which prints raw request headers and bodies directly rather than through the suppressed FastMCP loggers.

Impact: Submitted text, PII, compartment identifiers, and Signature authorization data can enter logs. In stdio mode, the output also corrupts the JSON-RPC stream.

Requested change: Force `log_requests=False` for every server-created OCI client and add a sentinel-payload test proving request content and authorization values never reach stdout or stderr.

File: src/oci-language-mcp-server/Containerfile [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  chown -R oracle:oracle /app /common
+
+USER oracle
+
+ENV LANGUAGE_MCP_HOST=0.0.0.0 \
+    LANGUAGE_MCP_PORT=8080 \
+    LANGUAGE_MCP_LOG_LEVEL=INFO \
+    LANGUAGE_MCP_LOG_FORMAT=json \
+    LANGUAGE_MCP_DEPLOYMENT_MODE=local \
+    LANGUAGE_MCP_TRANSPORT=streamable-http \
```
Comment (dustin-sale) [inline src/oci-language-mcp-server/Containerfile]: _Comment from @dustin-sale via Codex._

[P1] Do not expose unauthenticated local mode on all interfaces

Evidence: The default image binds Streamable HTTP to `0.0.0.0` while selecting `DEPLOYMENT_MODE=local`. Local mode performs no caller authentication, Origin is optional, and a network client can supply an allowed `Host: localhost` header.

Impact: Shared container networks, automatic port publishing, or an orchestrator Service can expose OCI Language operations under the container's configured identity.

Requested change: Default the image to stdio, or reject non-loopback local-mode listeners and require authenticated remote mode before binding to `0.0.0.0`. Add a default-image configuration regression test.

File: src/oci-language-mcp-server/oracle/oci_language_mcp_server/http_security.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+        path = str(scope.get("path", ""))
+        if path.startswith("/mcp") and self.settings.deployment_mode == "remote":
+            authorization = headers.get("authorization", "")
+            if self.settings.http_auth_mode == "token-file":
+                expected = f"Bearer {self.remote_token}"
+                if not hmac.compare_digest(authorization, expected):
+                    await self._reject(scope, receive, send, 401, "Authentication required.")
+                    return
+            if authorization and not await self._within_rate_limit():
```
Comment (dustin-sale) [inline src/oci-language-mcp-server/oracle/oci_language_mcp_server/http_security.py]: _Comment from @dustin-sale via Codex._

[P1] Authenticate before charging the OAuth quota

Evidence: In OAuth mode, any nonempty Authorization header reaches the global rate-limit deque before FastMCP validates the JWT. A local probe with a limit of one returned 401 for the first garbage bearer token and 429 for the second.

Impact: An unauthenticated caller can continuously consume the process-wide quota and deny service to legitimate OAuth clients.

Requested change: Charge the authenticated quota only after successful verification, preferably per principal, and keep pre-auth abuse limiting separate. Add a test proving invalid-token floods cannot block a subsequent valid request.

File: src/oci-language-mcp-server/Containerfile [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+      io.modelcontextprotocol.server.name="io.github.oracle/oci-language-mcp"
+
+RUN microdnf install epel-release && \
+  microdnf install python3.13 python3.13-pip \
+  && rm -rf /var/cache/dnf/* \
+  && useradd -m -d /app oracle
+
+WORKDIR /app
+COPY --chown=oracle:oracle src/common /common
+COPY --chown=oracle:oracle src/oci-language-mcp-server /app
```
Comment (dustin-sale) [inline src/oci-language-mcp-server/Containerfile]: _Comment from @dustin-sale via Codex._

[P1] Use one compatible container build context

Evidence: These paths require repository-root context, but `make containerize project=oci-language-mcp-server` changes into the server directory before building, where neither source path exists. The documented root-context Docker build instead ignores the nested `.containerignore`, because no ignore file exists at the context root.

Impact: The standard repository target fails, while the documented alternative can copy local virtual environments, coverage artifacts, or secret files into image layers.

Requested change: Establish one supported context for Make and README, install effective context-root ignore rules, narrow the copied paths, and validate the exact repository container target in CI.

File: src/oci-language-mcp-server/oracle/oci_language_mcp_server/service.py [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+                request_id=request_id,
+                client_opc_request_id=client_opc_request_id,
+                submitted=submitted,
+                code=code,
+                message=message,
+                retryable=retryable,
+                oci_request_id=extract_service_error_oci_request_id(exc),
+            )
+        except Exception as exc:
+            timed_out = type(exc).__name__ in {"ReadTimeout", "ConnectTimeout"}
```
Comment (dustin-sale) [inline src/oci-language-mcp-server/oracle/oci_language_mcp_server/service.py]: _Comment from @dustin-sale via Codex._

[P2] Classify OCI read timeouts as retryable

Evidence: Timeout detection checks exact class names, but OCI SDK 2.179.0 wraps read timeouts in `oci.exceptions.RequestException`; it exposes no public `ReadTimeout` exception here.

Impact: A configured read timeout becomes `INTERNAL_ERROR` with `retryable=false`, discouraging recovery from a transient failure.

Requested change: Classify OCI request wrappers and circuit-breaker failures using their actual types and safe cause chains, then add wrapped-read-timeout and open-circuit tests.

File: src/oci-language-mcp-server/oracle/oci_language_mcp_server/parser.py [hunk 7]
```diff
@@ -0,0 +1,560 @@
...
+    request: AnyToolRequest,
+    request_id: str,
+    client_opc_request_id: str,
+) -> AnyToolResult:
+    """Convert an OCI SDK response without leaking raw upstream errors."""
+
+    submitted = len(request.documents)
+    data = getattr(response, "data", response)
+    payload = oci.util.to_dict(data) if data is not None else {}
+    documents = [_parse_document(tool, item, request) for item in payload.get("documents") or []]
```
Comment (dustin-sale) [inline src/oci-language-mcp-server/oracle/oci_language_mcp_server/parser.py]: _Comment from @dustin-sale via Codex._

[P2] Reconcile response keys with the submitted batch

Evidence: Parsed documents and errors are counted without checking their keys against the request. A duplicate-response probe produced `submitted=1`, `succeeded=2`, and overall `status=succeeded`; missing and unknown keys are likewise accepted.

Impact: Partial or malformed OCI responses can silently lose, duplicate, or misattribute document results.

Requested change: Validate all returned document/error keys, reject duplicates and unknowns, synthesize sanitized errors for missing keys, and add malformed-response regressions.

File: src/oci-language-mcp-server/oracle/oci_language_mcp_server/parser.py [hunk 8]
```diff
@@ -0,0 +1,560 @@
...
+        assert isinstance(request, DetectPiiEntitiesRequest)
+        include_text = request.masking is None or request.options.include_original_entity_text
+        masked_text = _value(item, "masked_text", "maskedText")
+        if request.masking and any(
+            isinstance(rule, RemoveRule) for rule in request.masking.values()
+        ):
+            masked_text = (
+                _normalize_removed_text(masked_text)
+                if isinstance(masked_text, str)
+                else masked_text
```
Comment (dustin-sale) [inline src/oci-language-mcp-server/oracle/oci_language_mcp_server/parser.py]: _Comment from @dustin-sale via Codex._

[P2] Preserve non-PII formatting in REMOVE results

Evidence: Whenever a REMOVE rule is present, `_normalize_removed_text` collapses spaces and tabs, strips indentation, and changes punctuation spacing throughout the entire document, not only around removed entities.

Impact: Tables, code, and other formatting-sensitive non-PII content can be corrupted.

Requested change: Return OCI's `masked_text` unchanged and add a regression containing unrelated indentation, repeated spaces, and tabs.

File: src/oci-language-mcp-server/oracle/oci_language_mcp_server/models.py [hunk 9]
```diff
@@ -0,0 +1,532 @@
...
+
+MaskingRule = Annotated[
+    MaskRule | ReplaceRule | RemoveRule | RelexifyRule,
+    Field(discriminator="mode"),
+]
+
+
+class DetectPiiEntitiesRequest(BatchRequest):
+    documents: list[EnglishDocument] = Field(min_length=1, max_length=MAX_DOCUMENTS)
+    masking: dict[str, MaskingRule] | None = Field(
```
Comment (dustin-sale) [inline src/oci-language-mcp-server/oracle/oci_language_mcp_server/models.py:283]: _Comment from @dustin-sale via Codex._

[P2] Define consistent semantics for an empty masking map

Evidence: `masking={}` passes validation. It is treated as non-transformed by truthiness checks, but the parser uses `request.masking is None`, causing detected entity text to be omitted.

Impact: A successful detection-only request can report PII entities without their detected values.

Requested change: Reject empty maps, normalize them to `None`, or use consistent truthiness semantics. Test omitted, null, and empty masking values.

File: src/oci-language-mcp-server/Containerfile [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+    LANGUAGE_MCP_LOG_FORMAT=json \
+    LANGUAGE_MCP_DEPLOYMENT_MODE=local \
+    LANGUAGE_MCP_TRANSPORT=streamable-http \
+    UV_NO_SYNC=1
+
+EXPOSE 8080
+STOPSIGNAL SIGTERM
+
+HEALTHCHECK --interval=30s --timeout=5s --start-period=5s --retries=3 \
+    CMD python3.13 -c "import urllib.request; urllib.request.urlopen('http://localhost:8080/health')" || exit 1
```
Comment (dustin-sale) [inline src/oci-language-mcp-server/Containerfile]: _Comment from @dustin-sale via Codex._

[P2] Make the healthcheck deployment-aware

Evidence: Containerized stdio opens no HTTP listener. In documented remote deployments, `localhost` is generally absent from `ALLOWED_HOSTS`, so this probe is rejected before reaching `/health`.

Impact: Healthy stdio or remote containers can be marked unhealthy, triggering alerts or restarts.

Requested change: Make the healthcheck conditional on transport and provide a remote-mode probe whose Host value satisfies the configured policy, or document and configure an explicit runtime override.

File: src/oci-language-mcp-server/oracle/oci_language_mcp_server/models.py [hunk 11]
```diff
@@ -0,0 +1,532 @@
...
+
+class DetectDominantLanguageRequest(BatchRequest):
+    documents: list[PlainTextDocument] = Field(min_length=1, max_length=MAX_DOCUMENTS)
+    should_ignore_transliteration: bool = Field(
+        default=False,
+        description="Ignore transliterated text while identifying the language.",
+    )
+    chars_to_consider: int | None = Field(
+        default=None,
+        ge=1,
```
Comment (dustin-sale) [inline src/oci-language-mcp-server/oracle/oci_language_mcp_server/models.py]: _Comment from @dustin-sale via Codex._

[P3] Permit OCI's all-characters value

Evidence: OCI documents `chars_to_consider=0` as “use all characters,” while this constraint rejects zero. Omitting the field instead asks OCI to choose an optimum count, which is different behavior.

Impact: Clients cannot request a supported dominant-language mode.

Requested change: Change the lower bound to zero and add a provider/tool-contract test proving zero reaches the SDK request unchanged.

**Review by dustin-sale — CHANGES_REQUESTED** [review]
_Review by @dustin-sale via Codex._

## Requested changes
- `models.py:120` — [P1] Restrict caller-controlled OCI regions.
- `provider.py:159` — [P1] Disable OCI SDK request logging.
- `Containerfile:28` — [P1] Do not expose unauthenticated local mode on all interfaces.
- `http_security.py:64` — [P1] Prevent invalid OAuth tokens from consuming the authenticated quota.
- `Containerfile:19` — [P1] Align the container build context with Make and effective ignore rules.

## Additional review notes
- Correct timeout classification and response-key reconciliation.
- Preserve REMOVE output formatting and handle empty PII masking maps consistently.
- Make the container healthcheck deployment-aware.
- Permit OCI's documented `chars_to_consider=0` behavior.

## Validation
`make lint` passed. `make test project=oci-language-mcp-server` passed all 87 tests with 91.00% branch coverage. No container image was executed.

See the inline comments in this review for evidence, impact, and suggested remediation.

**Comment by antimverma** [comment]
@dustin-sale Addressed the latest comments, please check.

File: src/oci-language-mcp-server/Containerfile [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+RUN microdnf install epel-release && \
+  microdnf install python3.13 python3.13-pip \
+  && rm -rf /var/cache/dnf/* \
+  && useradd -m -d /app oracle
+
+WORKDIR /app
+COPY --chown=oracle:oracle . /app
+
+RUN pip3.13 install --no-cache-dir uv && \
+  uv --no-cache sync --locked --all-extras
```
Comment (dustin-sale) [inline src/oci-language-mcp-server/Containerfile]: _Comment from @dustin-sale via Codex._

[P1] Disable development workspace sources in the production sync

Evidence: The server-only build context is appropriate, but `uv sync --locked` follows the development-only `tool.uv.sources` override and the lockfile's editable `../common` source. Reproducing this context fails because that workspace member is unavailable. Running `uv sync --no-sources` instead disables the workspace override and resolves the standard `oracle-mcp-common>=0.1.2,<0.2.0` dependency from PyPI.

Impact: The documented container build and `SUBDIRS=src/oci-language-mcp-server make containerize` currently fail before producing the advertised image.

Requested change: Keep the server-only context and existing development lockfile, but replace the locked workspace sync with `uv --no-cache sync --no-sources --no-dev --no-editable`. This retains `uv sync`, installs `oracle-mcp-common` from PyPI, omits development dependencies, and installs the server in deployment form. Validate the exact container build command in CI.

File: src/oci-language-mcp-server/README.md [hunk 13]
```diff
@@ -0,0 +1,313 @@
...
+reads the standard OCI config and security-token files; credentials are never configured as
+MCP arguments.
+
+## Containerized stdio and secured Streamable HTTP
+
+From the server directory, build the image:
+
+```bash
+cd src/oci-language-mcp-server
+docker build -t oci-language-mcp:0.1.0 .
```
Comment (dustin-sale) [inline src/oci-language-mcp-server/README.md]: _Comment from @dustin-sale via Codex._

[P2] Follow the repository Podman workflow

Evidence: The root project README defines Podman as the supported container workflow and builds server images from the repository root with `SUBDIRS=src/<server> make containerize`. That target enters the server directory, runs `podman build`, and creates both the package-version and `latest` tags. This server README instead documents direct `docker build` and `docker run` commands with a different image name.

Impact: Users bypass the repository’s standard build target, tagging convention, and `.containerignore` behavior, and the server documentation can drift from the container path exercised by the project.

Requested change: Follow the root README by documenting `SUBDIRS=src/oci-language-mcp-server make containerize` from the repository root. Replace the Docker run examples with `podman run`, use the generated `oracle.oci-language-mcp-server:latest` or versioned image tag, and keep the mounted bearer-token file outside the server build context.

**Review by dustin-sale — CHANGES_REQUESTED** [review]
_Review by @dustin-sale via Codex._

## Requested changes
- `Containerfile:22` — [P1] Disable development workspace sources when syncing the production image so `oracle-mcp-common` installs from PyPI.
- `README.md:101` — [P2] Follow the repository’s Podman-based container build and run workflow.

## Re-review result
Ten previous findings are addressed: caller-controlled region validation, OCI request logging, unauthenticated container defaults, OAuth quota placement, timeout classification, response-key reconciliation, REMOVE formatting, empty masking semantics, healthcheck behavior, and `chars_to_consider=0`.

The remaining issues are limited to container packaging. The server-only context is appropriate, but the current locked sync still requires the absent development workspace, and the server README diverges from the root project’s supported Podman workflow.

## Validation
`make lint` passed. `make test project=oci-language-mcp-server` passed all 95 tests with 91.11% branch coverage. Package lock and build checks passed with the repository workspace present. A focused OAuth probe confirmed invalid tokens do not consume the authenticated quota.

In an isolated server-only context, the current `uv sync --locked` path failed on the unavailable `../common` workspace. Disabling development sources allowed the standard `oracle-mcp-common>=0.1.2,<0.2.0` dependency to resolve from PyPI while retaining the version already selected by the development lock.

**Comment by antimverma** [comment]
@dustin-sale Addressed the latest review comments, please check.

File: src/oci-language-mcp-server/README.md [hunk 14]
```diff
@@ -0,0 +1,312 @@
...
+Create or refresh a temporary local identity with `oci session authenticate`. Session mode
+reads the standard OCI config and security-token files; credentials are never configured as
+MCP arguments.
+
+## Containerized stdio and secured Streamable HTTP
+
+From the repository root, build the image with the supported Podman workflow:
+
+```bash
+SUBDIRS=src/oci-language-mcp-server make containerize
```
Comment (dustin-sale) [inline src/oci-language-mcp-server/README.md:100]: _Comment from @dustin-sale via Codex._

[P1] Make the documented container target self-contained

- Evidence: In a clean detached checkout bootstrapped from `requirements-dev.txt`, the exact documented command fails twice with `Failed to spawn: tomlq`. The shared `containerize` recipe invokes `uv run tomlq`, but neither repository requirements file declares that executable. The recipe then continues with empty image name and version values. The passing language-server CI job runs `uv build`, not this container target.

- Impact: Users following this README cannot produce the advertised image from the repository-supported workflow.

- Requested change: Make the shared `containerize` target resolve project name and version without an undeclared executable—for example, use Python `tomllib` as the repository’s other Make targets already do—and make metadata-resolution failures terminate the recipe. Add validation that exercises the exact documented target.

This issue already existed before this PR, but might as well fix it. Adding `tomlq` to `requirements-dev.txt` should resolve this issue.

File: src/oci-language-mcp-server/README.md [hunk 15]
```diff
@@ -0,0 +1,312 @@
...
+security controls. It is not an Oracle-hosted MCP endpoint. Each tool call invokes one OCI
+Language batch API and can incur service usage.
+
+See the official [OCI Language overview](https://docs.oracle.com/en-us/iaas/Content/language/using/overview.htm),
+[pretrained models](https://docs.oracle.com/en-us/iaas/Content/language/using/pretrain-models.htm),
+and [service limits](https://docs.oracle.com/en-us/iaas/Content/language/using/model-limitations.htm).
+
+## Requirements and IAM
+
+- Python 3.13 or Docker
```
Comment (dustin-sale) [inline src/oci-language-mcp-server/README.md]: _Comment from @dustin-sale via Codex._

[P2] List the container runtime actually used by the workflow

- Evidence: This requirement still says “Python 3.13 or Docker,” while the root project guidance and every container build/run command in this README require Podman.

- Impact: A user can satisfy the stated prerequisite yet still be unable to run the documented commands.

- Requested change: Replace Docker with Podman here, and refresh the PR description’s remaining Docker and repository-root-context wording.

**Review by dustin-sale — CHANGES_REQUESTED** [review]
_Review by @dustin-sale via Codex._

## Requested changes
- `src/oci-language-mcp-server/README.md:100` — [P1] Make the documented Podman build path self-contained.

## Additional review notes
- `src/oci-language-mcp-server/README.md:28` — [P2] List Podman, not Docker, as the container prerequisite.

## Validation
`make lint` and `make test project=oci-language-mcp-server` passed; all 95 tests passed with 91.11% branch coverage. The isolated production sync installed `oracle-mcp-common==0.1.2` from PyPI, but the documented `make containerize` command failed because `tomlq` is undeclared.

See the inline comments in this review for evidence, impact, and suggested remediation.

**Comment by antimverma** [comment]
@dustin-sale Addressed the latest review comments, please check.

File: src/oci-language-mcp-server/README.md (hunk 14 above)
Comment (dustin-sale) [reply src/oci-language-mcp-server/README.md:100]: My PR review skill is still not happy with `make containerize`. See my comment in `Makefile`.

File: Makefile [hunk 16]
```diff
@@ -192,8 +192,8 @@ e2e-tests: build install
 containerize:
 	@for dir in $(SUBDIRS); do \
 		if [[ -f $$dir/Containerfile && (-f $$dir/pyproject.toml) ]]; then \
-			name=$$(uv run tomlq -r '.project.name' $$dir/pyproject.toml); \
-			version=$$(uv run tomlq -r '.project.version' $$dir/pyproject.toml); \
+			name=$$(python3 -c "import sys, tomllib; print(tomllib.load(open(sys.argv[1], 'rb'))['project']['name'])" $$dir/pyproject.toml) || exit 1; \
```
Comment (dustin-sale) [inline Makefile]: My PR review skill still isn't happy with these lines, updating them to the following should resolve its issues with python3 not resolving to 3.13.

> name=$$(uv run --isolated --no-project --python 3.13 python -c "import sys, tomllib; print(tomllib.load(open(sys.argv[1], 'rb'))['project']['name'])" "$$dir/pyproject.toml") || exit 1; \
> version=$$(uv run --isolated --no-project --python 3.13 python -c "import sys, tomllib; print(tomllib.load(open(sys.argv[1], 'rb'))['project']['version'])" "$$dir/pyproject.toml") || exit 1; \

**Comment by antimverma** [comment]
Hi @dustin-sale , Gentle Reminder to review this ^^

**Review by dustin-sale — CHANGES_REQUESTED** [review]
Really close, my PR review skill isn't happy with the containerize make step. I think my suggestion should resolve it.

**Comment by antimverma** [comment]
> Really close, my PR review skill isn't happy with the containerize make step. I think my suggestion should resolve it.

@dustin-sale Pushed the change.

**Review by dustin-sale — APPROVED** [review]

File: src/oci-language-mcp-server/oracle/oci_language_mcp_server/__main__.py [hunk 17]
```diff
@@ -0,0 +1,12 @@
...
+# Licensed under the Universal Permissive License v1.0 as shown at
+# https://oss.oracle.com/licenses/upl.
+
+"""Run the OCI Language MCP server."""
+
+from __future__ import annotations
+
+from .server import main
+
+if __name__ == "__main__":
```
Comment (gebhardtr) [inline src/oci-language-mcp-server/oracle/oci_language_mcp_server/__main__.py]: do you need this file, given that `server` is the entrypoint (and has this same main check)?

**Review by gebhardtr — CHANGES_REQUESTED** [review]
comment re: 2nd entry point. https://github.com/oracle/mcp/pull/406#pullrequestreview-4932456721

File: src/oci-language-mcp-server/oracle/oci_language_mcp_server/__main__.py (hunk 17 above)
Comment (antimverma) [reply src/oci-language-mcp-server/oracle/oci_language_mcp_server/__main__.py]: @gebhardtr Yes good catch, this was a redundant residual code. Addressed.

File: src/oci-language-mcp-server/oracle/oci_language_mcp_server/provider.py [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+    @staticmethod
+    def _call_options(opc_request_id: str) -> dict[str, Any]:
+        return {
+            "opc_request_id": opc_request_id,
+            "retry_strategy": oci.retry.NoneRetryStrategy(),
+        }
+
+    def _client(self, *, region: str | None) -> Any:
+        auth = build_auth_context(self._settings, region=region)
+        cache_key = (auth.mode, auth.region, self._settings.oci_service_endpoint)
```
Comment (gebhardtr) [inline src/oci-language-mcp-server/oracle/oci_language_mcp_server/provider.py]: is a caching layer necessary? why are we managing multiple clients? Hit from codex:

> [P2] Synchronize the regional client cache. [LanguageService](https://github.com/oracle/mcp/blob/8fdf84bf7349ab72faf2567c27753af9cbcb92f9/src/oci-language-mcp-server/oracle/oci_language_mcp_server/service.py#L63-L65) invokes the shared provider from up to eight threads, while [_client()](https://github.com/oracle/mcp/blob/8fdf84bf7349ab72faf2567c27753af9cbcb92f9/src/oci-language-mcp-server/oracle/oci_language_mcp_server/provider.py#L156-L182) performs compound, unsynchronized operations on a shared OrderedDict. During regional cache churn, another worker can evict a key between the membership check and move_to_end, producing a KeyError and an INTERNAL_ERROR; concurrent misses can also create duplicate clients. Protect lookup/insertion/eviction with a lock and add a concurrent churn test.

Comment (antimverma) [reply src/oci-language-mcp-server/oracle/oci_language_mcp_server/provider.py]: This is addressed.

**Comment by antimverma** [comment]
@gebhardtr All comments addressed. Can you please review.
